File: ext/qed/channels/imaginary.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2022 all rights reserved


// external
#include "external.h"
// namespace setup
#include "forward.h"
// the imaginary channel
#include "imaginary.h"


// imaginary
void
qed::py::channels::imaginary(py::module & m)
{
    // bindings for {mapgrid_t} sources
    // add the bindings for {c8}
    m.def(
        // the name of the function
        "imaginaryComplexFloatGrid",
        // the handler
        &imaginaryGridTile<mapgrid_t<std::complex<float>>>,
        // the signature
        "source"_a, "zoom"_a, "origin"_a, "shape"_a, "min"_a, "max"_a,
        // the docstring
        "render the imaginary part a complex float tile");
    // and the bindings for {c16}
    m.def(
        // the name of the function
        "imaginaryComplexDoubleGrid",
        // the handler
        &imaginaryGridTile<mapgrid_t<std::complex<double>>>,
        // the signature
        "source"_a, "zoom"_a, "origin"_a, "shape"_a, "min"_a, "max"_a,
        // the docstring
        "render the imaginary part of a complex double tile");

    // bindings for HDF5 sources
    // add the bindings for {c8}
    m.def(
        // the name of the function
        "imaginaryComplexFloatHDF5",
        // the handler
        &imaginaryHDF5Tile<heapgrid_t<std::complex<float>>>,
        // the signature
        "source"_a, "zoom"_a, "origin"_a, "shape"_a, "min"_a, "max"_a,
        // the docstring
        "render the imaginary part a complex float tile");
    // and the bindings for {c16}
    m.def(
        // the name of the function
        "imaginaryComplexDoubleHDF5",
        // the handler
        &imaginaryHDF5Tile<heapgrid_t<std::complex<double>>>,
        // the signature
        "source"_a, "zoom"_a, "origin"_a, "shape"_a, "min"_a, "max"_a,
        // the docstring
        "render the imaginary part of a complex double tile");

    // all done
    return;
}


// end of file

File: pkg/readers/isce2/interferogram/channels/Imaginary.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis <michael.aivazis@para-sim.com>
# (c) 1998-2022 all rights reserved


# support
import qed
# superclass
from .Channel import Channel


# a channel for displaying the imaginary part of complex values
class Imaginary(Channel, family="qed.channels.isce2.int.imaginary"):
   """
   Make a visualization pipeline to display the imaginary part of complex values
   """


   # configurable state
   range = qed.protocols.controller(default=qed.controllers.linearRange)
   range.doc = "the manager of the range of values to render"


   # interface
   def autotune(self, **kwds):
      """
      Use the {stats} gathered on a data sample to adjust the range configuration
      """
      # chain up
      super().autotune(**kwds)
      # notify my range
      self.range.autotune(**kwds)
      # all done
      return


   def controllers(self, **kwds):
      """
      Generate the controllers that manipulate my state
      """
      # chain up
      yield from super().controllers(**kwds)
      # my range
      yield self.range, self.pyre_trait(alias="range")
      # all done
      return


   def eval(self, pixel):
      """
      Get the {pixel} value
      """
      # easy enough
      return pixel.imag


   def project(self, pixel):
      """
      Compute the imaginary part of a {pixel}
      """
      # easy
      yield pixel.imag, ""
      # all done
      return


   def tile(self, **kwds):
      """
      Generate a tile of the given characteristics
      """
      # add my configuration and chain up
      return super().tile(min=self.range.low, max=self.range.high, **kwds)


   # constants
   tag = "imaginary"


# end of file
